### DISCUSSION TITLE
How to log the action space variables to tensorboard

### DISCUSSION THREAD

**Discussion by maths4513** [Toni-SM/skrl#84] [Q&A] [answered]
I am able to run the skrl along with isaacgym env, everything is great.

I tried to log the variables from action space to tensorboar use  agent.track_data api, just like this:

```python
 98 agent = TRPO(models=models_trpo,
 99             memory=memory,
100             cfg=cfg_trpo,
101             observation_space=env.observation_space,
102             action_space=env.action_space,
103             device=device)
104 print(type(env.action_space))
**105 agent.track_data("target velocity", env.action_space[0])**
```

with the print information, it seems like the env.action_space inherent ed from isaacgym, and it is a gym.space.Box data type.

I am wondering is this the right way to log the scalar variables from action_space?

**Comment by Toni-SM** [comment, marked as answer]
Hi @maths4513 

If you want to log the actions taken by the policy (rather than the action space: definition of the action expected by the environment) currently you need to modify the agent or the trainer to include the `agent.track_data(...)` call   

Also, you can run the training manually and log the variables you want as follow:

```python
...
agent = TRPO(models=models_trpo,
             memory=memory,
             cfg=cfg_trpo,
             observation_space=env.observation_space,
             action_space=env.action_space,
             device=device)


# Run training manually
# initialize agent
agent.init()

# reset environment
states, infos = env.reset()

import tqdm
timesteps = 1600

for timestep in tqdm.tqdm(range(timesteps)):
    # pre-interaction
    agent.pre_interaction(timestep=timestep, timesteps=timesteps)

    # compute actions
    with torch.no_grad():
        actions = agent.act(states, timestep=timestep, timesteps=timesteps)[0]

        # track actions
        agent.track_data("Actions / Action (max)", torch.max(actions).item())
        agent.track_data("Actions / Action (min)", torch.min(actions).item())
        agent.track_data("Actions / Action (mean)", torch.mean(actions).item())

    # step the environments
    next_states, rewards, terminated, truncated, infos = env.step(actions)

    # record the environments' transitions
    with torch.no_grad():
        agent.record_transition(states=states,
                                actions=actions,
                                rewards=rewards,
                                next_states=next_states,
                                terminated=terminated,
                                truncated=truncated,
                                infos=infos,
                                timestep=timestep,
                                timesteps=timesteps)

    # post-interaction
    agent.post_interaction(timestep=timestep, timesteps=timesteps)

    # update states (environments are reset internally)
    states = next_states
```
<br>
This will generate the following Tensorboard cards (example for Isaac Gym preview 4 Cartpole environment with PPO):

![Screenshot from 2023-06-26 10-52-54](https://github.com/Toni-SM/skrl/assets/22400377/3a098341-268e-4021-966e-7ee7fa77a1cd)

**Reply by GF-Huang** [reply]
How about now? Is it still necessary to go through such complexity just to log observations or actions?

My suggestion is to add some hook functions to the configuration (e.g., `TrainerCfg`) that are called once per `step` by the agent or trainer, passing relevant information—such as observations and actions—that users might want to track, allowing users to decide precisely what to track.

**Comment by maths4513** [comment]
Thanks, it works, great!.
